modules/tensor_mechanics/src/actions/TensorMechanicsAction.C

```diff
@@ -44,87 +102,182 @@ TensorMechanicsAction::TensorMechanicsAction(const InputParameters & params) :
 
   if (_diag_save_in.size() != 0 && _diag_save_in.size() != _ndisp)
     mooseError("Number of diag_save_in variables should equal to the number of displacement variables " << _ndisp);
+
+  // plane strain consistency check
+  if (_out_of_plane != OutOfPlane::None && _ndisp != 2)
+    mooseError("Plane strain only works in 2 dimensions");
+
+  if (_out_of_plane != OutOfPlane::None)
+    mooseError("Not implemented");
 }
 
 void
 TensorMechanicsAction::act()
 {
-  // get a list of all subdomains first
-  const auto & subdomain_set = _problem->mesh().meshSubdomains();
-  std::vector<SubdomainID> subdomains(subdomain_set.begin(), subdomain_set.end());
+  //
+  // consistency check for the coordinate system
+  //
+
+  // get subdomain IDs
+  std::set<SubdomainID> subdomain_ids;
+  for (auto & name : _subdomain_names)
+    subdomain_ids.insert(_mesh->getSubdomainID(name));
+
+  // use either block restriction list or list of all subdomains in the mesh
+  const auto & check_subdomains = subdomain_ids.empty() ? _problem->mesh().meshSubdomains() : subdomain_ids;
 
   // make sure all subdomains are using the same coordinate system
-  _coord_system = _problem->getCoordSystem(subdomains[0]);
-  for (auto s : subdomains)
-    if (_problem->getCoordSystem(s) != _coord_system)
+  _coord_system = _problem->getCoordSystem(*check_subdomains.begin());
+  for (auto subdomain : check_subdomains)
+    if (_problem->getCoordSystem(subdomain) != _coord_system)
       mooseError("The TensorMechanics action requires all subdomains to have the same coordinate system");
 
-  auto tensor_kernel_type = getKernelType();
-  auto params = getParameters(tensor_kernel_type);
+  //
+  // Meta action which optionally spawns other actions
+  //
+  if (_current_task == "meta_action")
+  {
+    const std::string type = "GeneralizedPlaneStrainAction";
+    auto params = _action_factory.getValidParams(type);
+    // auto action = MooseSharedNamespace::static_pointer_cast<MooseObjectAction>(_action_factory.create("type", name() + "_gps", params));
```

Comment:
Yeah, I can remove that. The action is not completely done and this is stuff to create a dependent action.